### PR TITLE
Initial efforts for PHP8 support

### PR DESCRIPTION
**Pull request by peteraba**
There are two main topics to be solved:
1. Some PDO adapters are not compatible with \PDO classes they extend (Connection, Statement)
2. Reflection usage. Especially relying on the now deprecated [ReflectionParameter::getClass](https://www.php.net/manual/en/reflectionparameter.getclass.php). Additionally it seems like something has changed in regards to tokenization / autoloader, because the `ReflectionClass` constructor fails for the parsed migrations in `FileMigrationFinder`

### CODE REVIEW

**Comment by peteraba** [comment]
Furthermore it seems that the oldest PHPUnit version which supports PHP8 does not support PHP7.1.

@davidbyoung what do you think about a new major version?

**Comment by peteraba** [comment]
Given that 7.3 is nearing it's support, I'd probably go straight to 7.4, but up to you of course.

**Comment by davidbyoung** [comment]
Wow, thanks @peteraba!  I am fine with supporting 7.3 and up (including 8).  Same goes for PHPUnit - let's upgrade to the lowest major version that supports 7.3 and up.

File: src/Opulence/Framework/Tests/Console/Testing/PhpUnit/IntegrationTestCaseTest.php [hunk 1]
```diff
@@ -62,9 +62,9 @@ public function testCallingCommandWithMultiplePrompts()
         $this->testCase->execute('multipleprompts', [], [], ['foo', 'bar'])
             ->getResponseAssertions()
             ->outputEquals('Custom1Custom2');
-        $this->testCase->execute('multipleprompts', [], [], ['default1', 'default2'])
```
Comment (peteraba) [inline src/Opulence/Framework/Tests/Console/Testing/PhpUnit/IntegrationTestCaseTest.php:65]: @davidbyoung this is the only test I couldn't really fix, I'm not sure what it's trying to test to be honest.

**Comment by peteraba** [comment]
Still to fix:
- `Opulence\Framework\Tests\Console\Testing\PhpUnit\Mocks\IntegrationTestCaseTest::testCallingCommandWithMultiplePrompts`

**Comment by peteraba** [comment]
Okay, so I'm ready to call this ready for review, though I acknowledge that at least `IntegrationTestCaseTest::testCallingCommandWithMultiplePrompts` will need to be fixed.

What changed:
1. Upgrade of PHPUnit was required older versions do not support PHP 8. 
2. PHPUnit upgrade means that all (?) `setUp` methods had to be changed to have a `void` return value defined.
3. PHPUnit upgrade suggests that all `at` calls are refactored as it now raises a deprecation error. Technically it could be suppressed, but the `at` method is fragile anyway, so I think it was better to refactor.
4. PHP8 changes the tokenization of namespaces, therefore that had to be changed and an additional optional constant declaration was needed.
5. `ReflectionParameter::getClass` is also deprecated now, therefore a special utility was added to achieve the same result with using supported methods instead. (This could potentially be improved, but let's see what the review finds...)

File: src/Opulence/Framework/Tests/Console/Testing/PhpUnit/IntegrationTestCaseTest.php (hunk 1 above)
Comment (davidbyoung) [reply src/Opulence/Framework/Tests/Console/Testing/PhpUnit/IntegrationTestCaseTest.php:65]: I'm not really sure, either.  We'll leave it off for now.

**Review by davidbyoung — APPROVED** [review]

**Comment by davidbyoung** [comment]
This is one hell of a lot of work!  I cannot thank you enough for undertaking it.  I am going to rebase this onto a new branch for 1.2 since this will introduce some BC breaks.

**Comment by peteraba** [comment]
@davidbyoung, thank you! also for the cleanup in the `composer.json` files in the submodules.

**Pull request merged**